spec/spec_helper.rb

```diff
@@ -143,6 +143,9 @@ def set_hieradata(hieradata)
     FileUtils.rm_rf(@spec_global_env_temp)
     @spec_global_env_temp = nil
   end
+  c.after(:suite) do
+    RSpec::Puppet::Coverage.report!(100)
```

Comment:
It may seem counterintuitive at first, but **100% rspec-puppet spec test coverage should not be a goal.** 

For technical _and_ practical reasons, 100% coverage tends to become a significant development antipattern.  This problem is  particularly severe with rspec-puppet, which calculates coverage based on a the assumption that every resource in the catalog should be referenced in a test at least once.  This is absurdly redundant for a declarative _(ish)_ language like Puppet, and is also impossible in some perfectly reasonable situations.  Considering the scale of the modules we maintain, it is critical to avoid this.

As a good rule of thumb: **the only critical thing to test with Puppet spec tests is _logic_** (e.g., template output,  branching, interpolation, functions, types, etc.,).  As a sanity check, we often test the basic presence of major classes in the default case, but that is not required, particularly if acceptance tests are present.  Testing the catalog beyond this is usually counterproductive:

* Attempting to provide 1:1 test-to-resource mapping is likely to result in a lot of brittle tests that aren't useful, but incur a significant maintenance cost:

  * As a simple example: If you include a class that simply defines file resources, writing a spec test for the class itself is sufficient to assume the files will exist as defined unless there was logic involved that might prevent it―either in the presence of the file (Puppet logic) content of the file (template logic).  
  * Testing for the existence of each file in the catalog tells you nothing you didn't already know, since that is what the class does.  This is essentially testing the behavior of the Puppet compiler, which is beyond the spec the class should be testing.  
  * However, any time a resource is renamed or refactored, the tests will now break and require additional maintenance.  

* The goal of a 100% coverage rate is sometimes impossible: Some modules contain internal logic to create new resources with titles that cannot be anticipated.

* Acceptance tests are far more effective at both verifying that the resources in catalog are correct and complete _enough_ to achieve the desired results than any catalog-based tests.

* There is a tendency to think that a number like "100%" is somehow a measure of quality.  However, it is possible to achieve 100% coverage without testing any logic at all.

This is why we dropped Coverage checks in the past and don't rely on code generators like https://github.com/nwops/puppet-retrospec to scaffold our spec tests.